src/parser/parser.c

```diff
@@ -5208,13 +5213,8 @@ static void parse_external_declaration(void)
 
 	position_t const *const pos = &ndeclaration->base.pos;
 	if (current_scope != file_scope) {
-		if (GNU_MODE) {
-			warningf(WARN_PEDANTIC, pos,
-			         "nested function %N is a GCC extension", ndeclaration);
-		} else {
-			errorf(pos,
-			       "nested function %N is a GCC extension", ndeclaration);
-		}
+		warningf(GNU_MODE? WARN_PEDANTIC: WARN_GNU_EXTENSION, pos,
+			"nested function %N is a GCC extension", ndeclaration);
```

Comment:
I am not quite sure if I correctly understood the problem with my current implementation. I am going to update it according to my understanding of your suggestion above and would be happy for feedback on it.

Have you meant something like that?
In this case the GNU_EXTENSION warning is always prefered but the WARN_PEDANTIC warning always prints as default instead if the GNU_EXTENSION warning is disabled.

```c
if (current_scope != file_scope) {
	bool printed_warning = false;
	if (!GNU_MODE) {
		printed_warning = warningf(WARN_GNU_EXTENSION, pos,
			"nested function %N is a GCC extension", ndeclaration);
	}
	if (!printed_warning) {
		warningf(WARN_PEDANTIC, pos,
			"nested function %N is a GCC extension", ndeclaration);
	}
}
```